### PR TITLE
internally change regions to be covariant

### LINKED ISSUE #103676: Internally change regions to be covariant
**THIS ISSUE IS ONLY ABOUT AN IMPLEMENTATION DETAIL OF RUSTC, IT DOES NOT AFFECT THE RUST LANGUAGE.**

Right now regions in rustc are considered to be *contravariant* wherever they are used:

https://github.com/rust-lang/rust/blob/cdd7afeaadf1c48eafb4dff4452439fa5d13a775/compiler/rustc_middle/src/ty/relate.rs#L447-L454

https://github.com/rust-lang/rust/blob/cdd7afeaadf1c48eafb4dff4452439fa5d13a775/compiler/rustc_middle/src/ty/relate.rs#L492-L497

This is confusing and different from outlives, as `'a subtype 'b` currently means `'b outlives 'a`. Whenever we switch between "outlives" and "subtyping" for regions, we have to swap the order, e.g.

https://github.com/rust-lang/rust/blob/cdd7afeaadf1c48eafb4dff4452439fa5d13a775/compiler/rustc_infer/src/infer/outlives/env.rs#L137-L148

We've previously discussed this in https://rust-lang.zulipchat.com/#narrow/stream/144729-t-types/topic/relating.20regions.20variance where we ended up agreeing that the variance should be changed to covariance for regions.

Doing so is non-trivial as my first attempt at this ended up missing at least one place resulting in hard to diagnose errors in core. I still want to see this fixed and would be up to review or mentor any work here. This requires some familiarity with the compiler (some knowledge about borrowck and/or the lexical regionck would be helpful) to correctly update relevant comments when doing this change.

*edit: we also have to update the `crate_variances` query for this to work. probably that was the issue I encountered*

### PR DESCRIPTION
**Pull request by aliemjay**
Surprisingly, we consider the reference type `&'a T` to be contravaraint in its lifetime parameter. This is confusing and conflicts with the documentation we have in the reference, rustnomicon, and rustc-dev-guide. This also arguably not the correct use of terminology since we can use `&'static u8` in a place where `&' a u8` is expected, this implies that `&'static u8 <: &' a u8` and consequently `'static <: ' a`, hence covariance. 

Because of this, when relating two types, we used to switch the argument positions in a confusing way:
`Subtype(&'a u8 <: &'b u8) => Subtype('b <: 'a) => Outlives('a: 'b) => RegionSubRegion('b <= 'a)`

The reason for the current behavior is probably that we wanted `Subtype('b <: 'a)` and `RegionSubRegion('b <= 'a)` to be equivalent, but I don' t think this is a good reason since these relations are sufficiently different in that the first is a relation in the subtyping lattice and is intrinsic to the type-systems, while the the second relation is an implementation detail of regionck.

This PR changes this behavior to use covariance, so.. 
`Subtype(&'a u8 <: &'b u8) => Subtype('a <: 'b) => Outlives('a: 'b) => RegionSubRegion('b <= 'a) `

Resolves #103676 

r? @lcnr

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/variance/constraints.rs [hunk 1]
```diff
@@ -254,9 +253,8 @@ impl<'a, 'tcx> ConstraintContext<'a, 'tcx> {
             }
 
             ty::Dynamic(data, r, _) => {
-                // The type `Foo<T+'a>` is contravariant w/r/t `'a`:
-                let contra = self.contravariant(variance);
-                self.add_constraints_from_region(current, r, contra);
+                // The type `Foo<T + 'a>` is covariant w/r/t `'a`:
```
Comment (lcnr) [inline compiler/rustc_hir_analysis/src/variance/constraints.rs:256]:
```suggestion
                // The type `dyn Foo<T + 'a>` is covariant w/r/t `'a`:
```

**Review by lcnr — COMMENTED** [review]
nice :+1: 

please also update it in the region solver though as we still flip the variance of regions there right now

**Comment by aliemjay** [comment]
I've updated the PR description.

> please also update it in the region solver though as we still flip the variance of regions there right now

I think that can be an issue for a separate PR. For now, `Subtype('a <: 'b) => Outlives('a : 'b) => RegionSubRegion('b <= 'a)` makes sense to land as-is, IMO.

File: compiler/rustc_hir_analysis/src/variance/constraints.rs (hunk 1 above)
Comment (aliemjay) [reply compiler/rustc_hir_analysis/src/variance/constraints.rs:256]:
```suggestion
                // The type `dyn Trait<T> +'a` is covariant w/r/t `'a`:
```

**Comment by lcnr** [comment]
might impact perf

@bors try

> Surprisingly, we consider the reference type `&'a T` to be contravaraint in its lifetime parameter. This is confusing and conflicts with the documentation we have in the reference, rustnomicon, and rustc-dev-guide. This also arguably not the correct use of terminology since we can use `&'static u8` in a place where `&' a u8` is expected, this implies that `&'static u8 <: &' a u8` and consequently `'static <: ' a`, hence covariance. 

I don't think `'static <: 'a` is a consequence of `&'static u8 <: &'a u8`. You can equally as well say `&'static u8 <: &'a u8` holds if `'a <: 'static` holds which is what references being contravariant over regions means :thinking: 

---

I am not too convinced that there is a meaningful difference between `RegionSubRegion` and *subtyping for regions*. By flipping the variance inside of the relation of regions instead of the type relation itself this is still pretty confusing. E.g. seeing `lub` being used in `glb` is a pretty big red flag that something wrong here '^^

If you don't intend to work on changing `RegionSubRegion` I am fine with landing this as is while keeping an issue open to also flip `RegionSubRegion` (or ideally, remove the whole concept of `Sub` from regionck and nll and instead always talk about *outlives*)


r=me after perf

**Comment by lcnr** [comment]
@rust-timer queue

**Comment by aliemjay** [comment]
> I don't think `'static <: 'a` is a consequence of `&'static u8 <: &'a u8`. You can equally as well say `&'static u8 <: &'a u8` holds if `'a <: 'static` holds which is what references being contravariant over regions means thinking

Agree! In retrospect, this was a rather ridiculous argument :sweat smile:. 

This however leaves us with only one justification for the change, and that is to bring the implementation in line with the documentation (see [the Reference] and [Rustnomicon], which agree that `'static <: 'a`.)

[Rustnomicon]: https://doc.rust-lang.org/nomicon/subtyping.html
[the Reference]: https://doc.rust-lang.org/reference/subtyping.html#variance

> If you don't intend to work on changing `RegionSubRegion` I am fine with landing this as is while keeping an issue open to also flip `RegionSubRegion` (or ideally, remove the whole concept of `Sub` from regionck and nll and instead always talk about _outlives_)

One complication is that even when removing the concept of `Sub` from regionck we would still have to use `glb` in `lub` and vice-versa, because I don't think one can argue againts `lub('static, 'a) = 'a`.

Since this is more complicated than I initially believed, I think a T-types FCP may be necessary.

**Comment by lcnr** [comment]
> One complication is that even when removing the concept of `Sub` from regionck we would still have to use `glb` in `lub` and vice-versa, because I don't think one can argue against `lub('static, 'a) = 'static`.

One can argue against anything if one is so inclined :3 See this previous discussion: https://rust-lang.zulipchat.com/#narrow/stream/144729-t-types/topic/relating.20regions.20variance/near/284076240

For covariant regions to make sense the idea is to think of regions as a set of aliases. This is apparently how polonius models regions ^^ there the `lub('static, 'a) = 'a` does in fact hold, as `'static` means there are no aliasing regions or sth.

Not exactly sure what exactly is represented by the concept of an alias but I don't think we need an MCP for this as we've already had a previous discussion with signoff from niko, but we could open if you prefer ^^

**Comment by aliemjay** [comment]
> Not exactly sure what exactly is represented by the concept of an alias but I don't think we need an MCP for this as we've already had a previous discussion with signoff from niko, but we could open if you prefer ^^

mmm, I think we can ignore the FCP.

**Comment by lcnr** [comment]
alright depending on your vibe, @aliemjay 

@bors delegate+

**Pull request merged**